Issue 191: Add non-Flink stream processing example.

claudiofahey opened the pull request:
These examples are intended to illustrate how exactly-once semantics can be achieved with Pravega. To make it clear to the reader, the entirety of the stream processor is defined in a single source file of under 500 lines. In particular, these illustrative examples do not use Apache Flink.

tkaitchuck commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/NonRecoverableSingleThreadedProcessor.java:
```diff
@@ -0,0 +1,163 @@
+ * and writing each output event to another Pravega stream.
+ *
+ * This runs only a single thread.
+ * Upon restart, it reprocesses the entire input stream and recreates the output stream.
```
See my above comment.

claudiofahey replied:
This file will be removed from the PR.

tkaitchuck commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/README.md:
```diff
@@ -0,0 +1,177 @@
+in a single source file of under 500 lines.
+In particular, these illustrative examples do not use Apache Flink.
+
+As one reads and understands this example exactly-once processor, they should
```
This paragraph is trying to say too many things at the same time.

claudiofahey replied:
I cleaned this up.

tkaitchuck commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/RecoverableMultithreadedProcessor.java:
```diff
@@ -0,0 +1,279 @@
+                streamManager.createStream(scope, outputStreamName, streamConfig);
+            }
+
+            // Create a reader group manager. It must remain open to allow manual checkpoints to work.
```
This comment is not accurate. Once a checkpoint is initiated all readers will get it.

claudiofahey replied:
If I close the ReaderGroupManager immediately after getReaderGroup, calls to readerGroup.getOnlineReaders and readerGroup.initiateCheckpoint will never return. Is this a bug?

tkaitchuck replied:
No, it's not a bug. But it doesn't make the comment accurate. You only need to keep the reader group manager open to find out when the checkpoint has completed, not for it to actually complete.

claudiofahey replied:
The comment is now "It must remain open so that readerGroup.initiateCheckpoint returns".

tkaitchuck commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/RecoverableMultithreadedProcessor.java:
```diff
@@ -0,0 +1,279 @@
+        /**
+         * Perform a checkpoint, wait for it to complete, and write the checkpoint to the state.
+         */
+        private void performCheckpoint() {
```
Similar to the other class, this isn't really providing any value over automatic checkpoints. All this additional code defiantly makes the example more complex. It may be worth having one example of manual checkpoints somewhere, but it should be called out as one, and it should show a context where it is apparent, how they can provide value beyond what the automatic ones do.

claudiofahey replied:
I will delete this entire file. However this comment applies equally to ExactlyOnceMultithreadedProcessor and I replied to it above.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,495 @@
+                    if (status == Transaction.Status.OPEN) {
+                        log.info("commitTransaction: committing {}", transaction.getTxnId());
+                        transaction.commit();
+                        // Note that commit may return before the transaction is committed.
```
I have observed the following when running the code:
```
commitTransaction: transaction 00000000-0000-0000-0000-000000000006 status is OPEN
commitTransaction: committing 00000000-0000-0000-0000-000000000006
// transaction.commit(); called and returned here
transaction 00000000-0000-0000-0000-000000000006 status is COMMITTING
transaction 00000000-0000-0000-0000-000000000006 status is COMMITTING
transaction 00000000-0000-0000-0000-000000000006 status is COMMITTED
```
Does Pravega guarantee that the transaction will not abort after commit() even if the status remains as COMMITTING for a few seconds? Is the time in COMMITTING the delay that is referred to before the data can be read? Hopefully this is the case. Perhaps we can update the documentation to explain what the COMMITTING status means and the guarantees around it.

tkaitchuck replied:
The only valid transition from COMMITTING is COMMITTED. It cannot fail once it has reached COMMITTING.

tkaitchuck replied:
I've created an issue here: https://github.com/pravega/pravega/issues/3534

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,495 @@
+                }
+            });
+
+            // Wait for commit of all transactions in the checkpoint.
```
That's great. See my previous comment.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,495 @@
+                            log.info("commitTransaction: committed {}", transaction.getTxnId());
+                            break;
+                        } else if (status == Transaction.Status.ABORTED) {
+                            throw new RuntimeException(new TxnFailedException());
```
That's great. See my previous comment.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,495 @@
+            });
+            try {
+                workerExecutor.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
+            } catch (InterruptedException e) {
```
I will fix this.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,495 @@
+         *
+         * @param checkpointName
+         */
+        private void commitTransactions(String checkpointName) {
```
I do not understand how incremental recovery is possible using the Pravega API. Can you provide a sketch?

Here is my understanding. Please correct me where I am wrong. 
1. With Pravega reader groups, there is no way for an individual reader to obtain it's current offset, as would be possible with Kafka for example. Only the caller of initiateCheckpoint can get the stream cut and it will only get it once all readers have handled the checkpoint.
2. If a transaction is committed before the stream cut can be persisted to storage, then the recovery procedure will not know where to start a new reader group. (Should the recovery procedure use the original reader group instead?)
3. Therefore, we must guarantee the following order: initiateCheckpoint -> workers flush transactions, master waits for all readers by getting the checkpoint future -> master persists stream cut -> commit transactions.
4. Another complication with incremental recovery is that a new reader may not be assigned the same segments as a previously failed reader. This makes it very difficult to recover a consistent snapshot of the state in a stateful processor.

elizabethbain replied:
Sorry. I find the use of transactions in the discussion confusing. I thought checkpoints and transactions were not necessarily related, but 2 different mechanisms.

claudiofahey replied:
I find it easiest to think about what is needed to recover from a restart of the processor. The reader group must be positioned at a stream cut. Any transactions flushed prior to the stream cut but not committed must be committed. The transactions are flushed when readers get their checkpoint so that these are exactly aligned.

tkaitchuck replied:
@claudiofahey I think you are making things complicated by focusing on the need to recreate the whole reader group. That is in fact undesirable as it means restarting all the workers because one of them failed. 

Instead let's frame the goal as follows: We're trying to show how to perform exactly once processing. So following a failure reading must resume from the point at which the output was written. 

When a reader fails, readerOffline() can be invoked passing the position the reader failed at and the reader group will redistribute those segments at that position of the remaining readers or any new ones that join the group. 

So if every time a processor commits a transaction it stores the last position object, you can always resume from the correct location. This of course requires you to store the position object. There is a way around this though. Because it is possible that the reader might have persisted nothing before it died it is valid to pass `null` as the position object to `readerOffline()` this will result in it assuming the reader processed up to the last checkpoint but not beyond it. So if the processor calls commit on the transaction when it's reader receives a checkpoint event, and creates a new transaction. Then in response to any failure the correct action would be invoke readerOffline passing null for the position.

claudiofahey replied:
Ok, I can change the code to reuse the same reader group. That will avoid the need for manual checkpoints.

However, I don't see how to safely avoid the need for a two-phase commit process. If I understand you correctly, you are suggesting:

```
eventRead = reader.readNextEvent(1000);
// Assume eventRead.isCheckpoint().
transaction.commit();
transaction = writer.beginTxn();
// What happens if application crashes here?
// Call readNextEvent to inform reader group that we have processed this checkpoint.
eventRead = reader.readNextEvent(1000);
```

But consider what happens if the application crashes after commit() but before the reader group is updated with readNextEvent.
Upon recovery, the reader group will resume at the previous checkpoint and duplicate records will be created.
To avoid this, we still need to use the two-phase commit process of flush, persist transaction IDs, commit, update reader group.
I may be able to do this with another Pravega stream instead of a shared file system but I don't see how to completely eliminate
this need.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,495 @@
+                }
+
+                // Create "latest" file that indicates the latest checkpoint name.
+                // This file must be updated atomically.
```
The checkpoint directory is created when the checkpoint is initiated. However it will not contain a complete checkpoint until the workers write their state and the master writes the stream cuts. The "latest" file is essentially a marker to indicate the latest checkpoint that completed. Should I create a "complete" marker file in the checkpoint directory to indicate this?

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,495 @@
+        }
+    }
+
+    public static class Worker implements Runnable {
```
Ok, I will split them.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,495 @@
+                            Path checkpointDirPath = checkpointRootPath.resolve(checkpointName);
+                            Path transactionIdFilePath = checkpointDirPath.resolve(CHECKPOINT_TRANSACTION_ID_FILE_NAME_PREFIX + this.workerIndex);
+
+                            // Must ensure that txnId is persisted to latest state before committing transaction!
```
Yes I believe I could do that but it is harder to understand why this would be safe to do. As it is now, only transactions that have been flushed and are ready to commit have their transaction IDs written to the files.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,495 @@
+                    Transaction<String> transaction = null;
+                    long eventCounter = 0;
+
+                    for (; ; ) {
```
I'm not understanding how I would replace this with an inner and outer loop. Can you provide a sketch?

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/NonRecoverableMultithreadedProcessor.java:
```diff
@@ -0,0 +1,209 @@
+ * and writing each output event to another Pravega stream.
+ *
+ * This supports multiple worker threads.
+ * Upon restart, it reprocesses the entire input stream and recreates the output stream.
```
Yes I agree that this whole file is unnecessary. This evolved into the exactly-once processor and is now obsolete. I will delete this file from the PR.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/NonRecoverableMultithreadedProcessor.java:
```diff
@@ -0,0 +1,209 @@
+                                Parameters.getMinNumSegments()))
+                        .build();
+                streamManager.createStream(scope, inputStreamName, streamConfig);
+                // Since we start reading the input stream from the earliest event, we must delete the output stream.
```
This file will be removed from the PR.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/NonRecoverableMultithreadedProcessor.java:
```diff
@@ -0,0 +1,209 @@
+            });
+            try {
+                workerExecutor.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
+            } catch (InterruptedException e) {
```
This file will be removed from the PR.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,495 @@
+
+            // Schedule periodic task to initiate checkpoints.
+            // If any execution of this task takes longer than its period, then subsequent executions may start late, but will not concurrently execute.
+            initiateCheckpointExecutor.scheduleAtFixedRate(this::performCheckpoint, checkpointPeriodMs, checkpointPeriodMs, TimeUnit.MILLISECONDS);
```
The performCheckpoint function is doing more than just calling readerGroup.initiateCheckpoint. It is also persisting the checkpoint object (stream cuts). How do I use automatic checkpoints to perform work after the checkpoint completes? Is there a callback I can use?

claudiofahey replied:
I removed the exactly-once processor from this PR.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,495 @@
+                            .build();
+                    streamManager.createStream(scope, inputStreamName, streamConfig);
+                    // Since we are starting processing from the beginning, delete and create a new output stream.
+                    // TODO: Should we truncate stream instead of deleting?
```
Ok I will change this.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,495 @@
+        }
+
+        /**
+         * Commit all transactions that are part of a checkpoint.
```
Ok I will change this.

claudiofahey replied:
Comment is now "Commit all transactions that have been opened since the last checkpoint."

claudiofahey replied:
I reverted this change. It is now back to "Commit all transactions that are part of a checkpoint." Transactions are associated with exactly one checkpoint. When a worker gets a checkpoint, it writes the TXID to the checkpoint directory (directory name matches the checkpoint name). This indicates that the TXID should be committed if and only if this particular checkpoint completes. If the checkpoint fails for any reason, the entire process is terminated and the next execution will resume from the last successful checkpoint.

claudiofahey replied:
I removed the exactly-once processor from this PR.

claudiofahey commented:
In the event that a single worker dies, the segments that it was reading will be automatically assigned to another worker. However, this code does not transfer the state between workers, leaving the state inconsistent. 

Also, from what point does the reader read the newly assigned segments? Does it begin from the last checkpoint?

vijikarthi commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventGenerator.java:
```diff
@@ -0,0 +1,98 @@
+                String routingKey = String.format("rk%02d", eventCounter % 10);
+                long intData = rand.nextInt(100);
+                sum += intData;
+                String generatedTimestampStr = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX").format(new Date());
```
dateformat can be instantiated outside the loop for performance reasons.

claudiofahey replied:
Fixed

vijikarthi commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,300 @@
+package io.pravega.example.streamprocessing;
```
Missing license info

claudiofahey replied:
fixed

vijikarthi commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,300 @@
+        // These files contain the Pravega transaction IDs that must be committed now.
+        Path checkpointDirPath = checkpointRootPath.resolve(checkpointName);
+        List<UUID> txnIds = IntStream
+                .range(0, numWorkers)
```
Will it not be an issue if the program is restarted with different number of worked thread count? Ideally the range should reflect as many number of transaction prefix file that we have in the checkpoint directory.

claudiofahey replied:
Yes this is a problem. I will fix this.

claudiofahey replied:
Fixed.

vijikarthi commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,300 @@
+            cleanCheckpointDirectory(checkpointDirPath);
+        } catch (final Exception e) {
+            log.warn("performCheckpoint: Error performing checkpoint", e);
+            // Ignore error. We will retry when we are scheduled again.
```
does it make sense to cleanup the checkpoint files in case of exception and commit any open transaction?

claudiofahey replied:
If any exception occurs, then during the next execution of performCheckpoint, all previous transactions will be committed and old checkpoint files will be deleted. If this commits open transactions persisting the checkpoint to disk, then we may not have exactly once guarantees. The only problem I see is if exceptions occur continuously. In this case, new checkpoint directories will be created and never deleted. This app could put a limit on failures and exit if it exceeds a threshold. Do you think I should do this?

claudiofahey replied:
Actually my comment above is incorrect. Transactions are associated with a single checkpoint. If that checkpoint fails, then all transactions associated with it need to be aborted and workers need to restart from the last successful checkpoint. The easiest way of doing this is to terminate and restart the application so the latest commit does this.

claudiofahey replied:
I removed the exactly-once processor from this PR.

vijikarthi commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessorWorker.java:
```diff
@@ -0,0 +1,137 @@
+package io.pravega.example.streamprocessing;
```
Missing license header

claudiofahey replied:
Fixed

vijikarthi commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessorWorker.java:
```diff
@@ -0,0 +1,137 @@
+public class ExactlyOnceMultithreadedProcessorWorker implements Callable<Void> {
+    private static final org.slf4j.Logger log = LoggerFactory.getLogger(ExactlyOnceMultithreadedProcessorWorker.class);
+
+    private static final String STATE_FILE_NAME_PREFIX = "state-worker-";
```
looks like unused variable?

claudiofahey replied:
Fixed

vijikarthi commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/NonRecoverableSingleThreadedProcessor.java:
```diff
@@ -0,0 +1,142 @@
+                    log.info("eventCounter={}, event={}",
+                            String.format("%06d", eventCounter),
+                            message);
+                    final CompletableFuture writeFuture = writer.writeEvent(routingKey, message);
```
We don't need to create a variable here to capture the result?

claudiofahey replied:
I wanted to make it clear to the reader that writeEvent returns a future and returning doesn't indicate that the event was actually written. I added a comment to the code to clarify this.

vijikarthi commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/Parameters.java:
```diff
@@ -0,0 +1,66 @@
+package io.pravega.example.streamprocessing;
```
missing license header

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventGenerator.java:
```diff
@@ -0,0 +1,98 @@
+            long sum = 0;
+            for (;;) {
+                eventCounter++;
+                String routingKey = String.format("rk%02d", eventCounter % 10);
```
Fixed routing key.

claudiofahey replied:
It is unclear to me how to make it better by doing it per-reader as opposed to per-readerGroup. Is the current method acceptable for this PR? Perhaps a future PR could improve this.

RaulGracia commented:
@claudiofahey due to issues we had in previous releases merging `develop` into `master`, we have changed the release approach and the current development branch is `dev`. We plan to delete `develop` to avoid confusions, so could you please re-open this PR against `dev`? Thanks!

claudiofahey commented:
This PR is ready for re-review.

tkaitchuck commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,323 @@
+ * See also {@link ExactlyOnceMultithreadedProcessorWorker}.
+ */
+
+public class ExactlyOnceMultithreadedProcessor implements Callable<Void> {
```
It would be good to add a bunch of comments here, at the top of the file and throughout explaining the overall design, what it's doing and why.

tkaitchuck commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java:
```diff
@@ -0,0 +1,323 @@
+     *
+     * @param checkpointName
+     */
+    private void commitTransactions(String checkpointName) {
```
It would be good to have a comment explaining the connection between the checkpoint and the transaction

claudiofahey replied:
I removed the exactly-once processor from this PR.

tkaitchuck commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/README.md:
```diff
@@ -0,0 +1,173 @@
+a Pravega application to maintain this mapping by itself.
+To achieve at-least-once semantics with Pravega, use the algorithms in this section.
+
+In the algorithms below, one process in a distributed system will be designated
```
Given that this is an example showing how to work without Flink, I see no reason this code should copy Flink's pattern. For applications that don't have the requirement which Flink does that they be able to rewind the group arbitrarily, it is much easier to eliminate the master process and just rely on automatic checkpoints. The recovery algorithm is simple, upon a checkpoint a worker commits their transaction, and stores their position. Upon failure, readerOffline is called passing the position.

claudiofahey replied:
The latest commit has at-least-once semantics and does indeed use only automatic checkpoints. It does not use transactions.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/AtLeastOnceProcessor.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public Void call() throws Exception {
+        // TODO: handle case when reader dies (if checkpoint timeout occurs, reset reader group to last successful checkpoint)
```
This case should now be handled in the latest commit.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/README.md:
```diff
@@ -0,0 +1,173 @@
+how exactly-once semantics can be achieved with Pravega.
+To make it clear to the reader, the entirety of the stream processor is defined
+within this package with minimal dependencies.
+In particular, these illustrative examples do *not* use Apache Flink.
```
Fixed.

claudiofahey commented:
The latest commit should provide at-least-once semantics in a clean way using only Pravega. I believe the code is complete but it does not have any automated tests yet. We also need to update the README to describe how this works.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventGenerator.java:
```diff
@@ -0,0 +1,99 @@
+                        String.format("%06d", eventCounter),
+                        String.format("%08d", sum),
+                        message);
+                final CompletableFuture writeFuture = writer.writeEvent(routingKey, message);
```
Added logging upon ack.

claudiofahey commented:
This PR is ready for review. For a description of this PR, see https://github.com/claudiofahey/pravega-samples/blob/issue-191-streamprocessing/pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/README.md.
Unit and integration tests can be run with `./gradlew pravega-client-examples:build`.

fpj commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ReaderGroupPruner.java:
```diff
@@ -0,0 +1,117 @@
+        // We must ensure that we add this reader to the membership synchronizer before the reader group.
+        membershipSynchronizer.startAsync();
+        membershipSynchronizer.awaitRunning();
+        task = executor.scheduleAtFixedRate(new PruneRunner(), heartbeatIntervalMillis, heartbeatIntervalMillis, TimeUnit.MILLISECONDS);
```
I'm wondering if it would be best to add some randomization to the period between runs of the prune runner. If distinct readers run it at the same period and close to each other, then we would have them calling reader offline unnecessarily for the same offline readers. 

I don't feel very strongly about the comment because:
- reader offline is idempotent
- crashes are rare
- it should only matter for a larger reader groups

In any case, if you feel that you want to introduce it here, then I think it is a small improvement.

claudiofahey replied:
I have randomized the initial delay and the period as you have suggested.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/AppConfiguration.java:
```diff
@@ -0,0 +1,88 @@
+        return getEnvVar("INSTANCE_ID", UUID.randomUUID().toString());
+    }
+
+    public String getStream1Name() {
```
AppConfiguration is shared by EventGenerator, AtLeastOnceApp, and EventDebugSink. Stream 1 is the output for EventGenerator and the input for AtLastOnceApp. Stream 2 is the output for AtLeastOnceApp and the input for EventDebugSink. To avoid the confusion from this point of view, I chose Stream 1 and Stream 2. It is a little odd, I admit.

tkaitchuck commented on pravega-client-examples/src/test/java/io/pravega/example/streamprocessing/WorkerProcess.java:
```diff
@@ -0,0 +1,136 @@
+                         EventWriterConfig.builder().build())) {
+
+                final AtLeastOnceProcessorInstrumented proc = new AtLeastOnceProcessorInstrumented(
+                        () -> ReaderGroupPruner.create(
```
Why does this take a function instead of the value?

claudiofahey replied:
It must take a supplier function for readerSupplier because the run method (https://github.com/pravega/pravega-samples/blob/f329c6b7efd17713998d085a5ac8a26f33b66a44/pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/AtLeastOnceProcessor.java#L55) needs to carefully create and close it.
As far as prunerSupplier, this could potentially be supplied directly by the caller as a value and the behavior will be correct.  Although not critical, by using a supplier, this can ensure that the pruner thread does not start until after the run() method starts. Additionally, I like the consistency in having the user provide suppliers for both readerSupplier and prunerSupplier.

tkaitchuck commented on pravega-client-examples/src/test/java/io/pravega/example/streamprocessing/WorkerProcessGroup.java:
```diff
@@ -0,0 +1,102 @@
+
+    /**
+     * Processors are guaranteed to not process events written after this method returns.
+     */
```
Is this intended to stay?

claudiofahey replied:
No. I have removed this commented code.

tkaitchuck commented on pravega-client-examples/src/test/java/io/pravega/utils/SetupUtils.java:
```diff
@@ -0,0 +1,353 @@
+/**
+ * Utility functions for creating an in-process Pravega server or connecting to an external Pravega server.
+ */
+@NotThreadSafe
```
This is sort of supprising given the class's name. Perhaps a different name should be used?

claudiofahey replied:
I borrowed this entire class from https://github.com/fpj/eo-ingestion/blob/master/src/test/java/io/pravega/utils/SetupUtils.java and made minimal changes. This class appears to be based on https://github.com/pravega/pravega/blob/master/test/integration/src/main/java/io/pravega/test/integration/utils/SetupUtils.java. All of these are annotated with `@NotThreadSafe`.

tkaitchuck replied:
Maybe this https://github.com/pravega/pravega/blob/master/test/integration/src/test/java/io/pravega/test/integration/PravegaResource.java would would be a better model. It takes advantage of the `@Rule` syntax to allow for single initialization without imposing thread safety issues.

tkaitchuck commented on pravega-client-examples/src/test/java/io/pravega/example/streamprocessing/WorkerProcessConfig.java:
```diff
@@ -0,0 +1,29 @@
+import io.pravega.client.ClientConfig;
+import lombok.Builder;
+
+@Builder
```
This is using lombok.

claudiofahey replied:
Yes, I now remember that I used lombok in quite a few places. I have now updated some of the try-with-resource blocks with lombok @Cleanup. There are still a few places that I left unchanged because I didn't want to change the behavior.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventDebugSink.java:
```diff
@@ -0,0 +1,98 @@
+        final ReaderGroupConfig readerGroupConfig = ReaderGroupConfig.builder()
+                .stream(Stream.of(getConfig().getScope(), getConfig().getStream2Name()))
+                .build();
+        try (ReaderGroupManager readerGroupManager = ReaderGroupManager.withScope(getConfig().getScope(), getConfig().getControllerURI())) {
```
There is a finally block for the inner try that uses the result of the outer try to delete the reader group. I don't think I can reduce this to one try block.

claudiofahey replied:
If you are referring to combining `try (StreamManager...)` and `try (ReaderGroupManager...)`, I separated these because the StreamManager can be closed before the long-lived main loop starts.

tkaitchuck replied:
I was referring to both the `ReaderGroupManager` and `EventStreamClientFactory`. You can have two resources associated with a block and it can also have a finally.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventGenerator.java:
```diff
@@ -0,0 +1,92 @@
+
+    public void run() throws Exception {
+        final ClientConfig clientConfig = ClientConfig.builder().controllerURI(getConfig().getControllerURI()).build();
+        try (StreamManager streamManager = StreamManager.create(getConfig().getControllerURI())) {
```
Same reason as above - I separated these because the StreamManager can be closed before the long-lived main loop starts.

tkaitchuck replied:
"Same here" referred to making this a private method.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/AtLeastOnceApp.java:
```diff
@@ -0,0 +1,163 @@
+                                new JSONSerializer<>(new TypeToken<SampleEvent>(){}.getType()),
+                                ReaderConfig.builder().build()),
+                        1000)
+                {
```
Done.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventDebugSink.java:
```diff
@@ -0,0 +1,98 @@
+
+    public void run() throws Exception {
+        final ClientConfig clientConfig = ClientConfig.builder().controllerURI(getConfig().getControllerURI()).build();
+        try (StreamManager streamManager = StreamManager.create(getConfig().getControllerURI())) {
```
Done.

tkaitchuck replied:
I don't see the change

RaulGracia commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventDebugSink.java:
```diff
@@ -0,0 +1,98 @@
+                            eventRead.getEvent());
+                }
+            }
+        }
```
Nit: please, place the finally keyword as `} finally {`, so we follow the same style as in the Pravega repo.

claudiofahey replied:
Fixed.

RaulGracia commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventDebugSink.java:
```diff
@@ -0,0 +1,98 @@
+        finally {
+            readerGroupManager.deleteReaderGroup(readerGroup);
+        }
+    }
```
Do we need to handle `addShutdownHook` here?

claudiofahey replied:
See my response about EventGenerator.java.

RaulGracia commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventGenerator.java:
```diff
@@ -0,0 +1,92 @@
+            Thread.sleep(1000);
+        }
+    }
+
```
Do we need to handle `addShutdownHook` here?

claudiofahey replied:
I don't think we need to handle this here. There is an `addShutdownHook` in `AtLeastOnceApp::run` that will gracefully shutdown (remove itself from the RG) when a user presses control-C or kills the process without `-9`. We don't need to do anything special to terminate the `EventGenerator` app.

RaulGracia commented on pravega-client-examples/src/test/java/io/pravega/example/streamprocessing/StreamProcessingTest.java:
```diff
@@ -0,0 +1,519 @@
+    @Test
+    public void noProcessorTest() throws Exception {
+        final String methodName = (new Object() {
+        }).getClass().getEnclosingMethod().getName();
```
Nit: can we put this in the same line?

claudiofahey replied:
Fixed.

RaulGracia commented on pravega-client-examples/src/test/java/io/pravega/utils/SetupUtils.java:
```diff
@@ -0,0 +1,353 @@
+                    .enableMetrics(false)
+                    .enableAuth(enableAuth)
+                    .enableTls(enableTls)
+                    //.certFile(getFileFromResource(CERT_FILE))   // pravega #2519
```
Do we need to clean this up?

claudiofahey replied:
Fixed. I removed all commented and unused code.

RaulGracia commented on pravega-client-examples/src/test/java/io/pravega/utils/SetupUtils.java:
```diff
@@ -0,0 +1,353 @@
+            log.info("Getting client config");
+            return ClientConfig.builder()
+                    .controllerURI(URI.create(inProcPravegaCluster.getControllerURI()))
+                    //.credentials(new DefaultCredentials(PRAVEGA_PASSWORD, PRAVEGA_USERNAME))
```
Same here.

claudiofahey replied:
Fixed.

RaulGracia commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/README.md:
````diff
@@ -0,0 +1,176 @@
+  
+  ```shell script
+  cd ~/pravega-samples
+  scripts/runAtLeastOnceApp.sh 2
````
This throws a compilation problem in class `SetupUtils`.

claudiofahey replied:
Fixed.

RaulGracia reviewed (changes requested):
Looks good overall, thanks for contributing this sample. I have few minor comments and one major: I have tested the sample and there is a compilation error in class `SetupUtils`. Once these are addressed, I will approve this PR.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventGenerator.java:
```diff
@@ -0,0 +1,92 @@
+                            getConfig().getMinNumSegments()))
+                    .build();
+            streamManager.createStream(getConfig().getScope(), getConfig().getStream1Name(), streamConfig);
+            streamManager.updateStream(getConfig().getScope(), getConfig().getStream1Name(), streamConfig);
```
I wanted to ensure that the stream scaling policy gets applied, even if the stream already exists. `createStream` will not change the scaling policy if the stream already exists, although this behavior is not mentioned in the Javadoc. I will add a comment to the code here to clarify.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventGenerator.java:
```diff
@@ -0,0 +1,92 @@
+        }
+
+        final Random rand = new Random(42);
+        @Cleanup final EventStreamClientFactory clientFactory = EventStreamClientFactory.withScope(getConfig().getScope(), clientConfig);
```
Done.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventGenerator.java:
```diff
@@ -0,0 +1,92 @@
+            log.info("{}", event);
+            final CompletableFuture<Void> writeFuture = writer.writeEvent(event.routingKey, event);
+            final long ackedSequenceNumber = sequenceNumber;
+            writeFuture.thenRun(() -> log.debug("Acknowledged: sequenceNumber={}", ackedSequenceNumber));
```
This makes the logs much easier to read at info level.

claudiofahey commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/AtLeastOnceApp.java:
```diff
@@ -0,0 +1,135 @@
+                .stream(Stream.of(getConfig().getScope(), getConfig().getStream1Name()))
+                .automaticCheckpointIntervalMillis(getConfig().getCheckpointPeriodMs())
+                .build();
+        @Cleanup ReaderGroupManager readerGroupManager = ReaderGroupManager.withScope(getConfig().getScope(), clientConfig);
```
Fixed.

claudiofahey commented:
@RaulGracia, I have addressed all of your comments. However, it appears that there is an intermittent timeout failure in the test kill5of6Test. I will continue to investigate.

RaulGracia commented:
@claudiofahey maybe the failure you are seeing are related to the sporadic failures we see in Pravega core repo related to tests involving Pravega standalone: https://github.com/pravega/pravega/issues/5864

RaulGracia commented:
@claudiofahey I tested the sample and it works. Also compiled with JDK11 and JDK8 and it also works (`./gradlew installDist`). But I also observed the failure in test `kill5of6Test`. Please, let's try to get this test passing so we can merge this one soon.

claudiofahey commented:
FYI, I ran the integration 20 times and all have passed:
```
for i in {1..20}; do ./gradlew clean test; done |& tee -a /tmp/test.log
grep "SUCCESSFUL" /tmp/test.log
BUILD SUCCESSFUL in 3m 42s
BUILD SUCCESSFUL in 3m 31s
BUILD SUCCESSFUL in 3m 56s
BUILD SUCCESSFUL in 3m 41s
BUILD SUCCESSFUL in 3m 47s
BUILD SUCCESSFUL in 6m 25s
BUILD SUCCESSFUL in 4m 18s
BUILD SUCCESSFUL in 4m 34s
BUILD SUCCESSFUL in 4m 25s
BUILD SUCCESSFUL in 4m 32s
BUILD SUCCESSFUL in 4m 16s
BUILD SUCCESSFUL in 4m 16s
BUILD SUCCESSFUL in 6m 29s
BUILD SUCCESSFUL in 4m 5s
BUILD SUCCESSFUL in 4m 8s
BUILD SUCCESSFUL in 4m 6s
BUILD SUCCESSFUL in 4m 26s
BUILD SUCCESSFUL in 4m 29s
BUILD SUCCESSFUL in 4m 2s
BUILD SUCCESSFUL in 3m 37s
```

RaulGracia commented:
@claudiofahey tests passed for me. I had to upgrade the Scala version of Spark in the Hadoop sample, as build was broken by one of the previous PRs. But now looks good, thanks!.

RaulGracia commented:
@tkaitchuck please, would you mind to do another review to either approve it or request more changes?

tkaitchuck commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/AtLeastOnceApp.java:
```diff
@@ -0,0 +1,140 @@
+        final ClientConfig clientConfig = ClientConfig.builder().controllerURI(getConfig().getControllerURI()).build();
+
+        // Create the input and output streams (ignored if they already exist).
+        try (StreamManager streamManager = StreamManager.create(clientConfig)) {
```
Let's make this a private method.

claudiofahey replied:
Done.

claudiofahey replied:
Done.

tkaitchuck commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/AtLeastOnceApp.java:
```diff
@@ -0,0 +1,140 @@
+        processor.startAsync();
+
+        // Add shutdown hook for graceful shutdown.
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
```
Shouldn't the processor do this itself if this is needed.?

claudiofahey replied:
It gets messy to do this in the AtLeastOnceProcessor because then it would need to deal with removing the shutdown hook, but only doing that when the JVM is not actually shutting down. The point of this shutdown hook is to shutdown the whole application gracefully, which in this case happens to consist of only one AtLeastOnceProcessor service. But in general, an application may consist of many services that all need to be shutdown in a particular order. So it seems appropriate that AtLeastOnceApp should coordinate the shutdown.

tkaitchuck commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventDebugSink.java:
```diff
@@ -0,0 +1,98 @@
+                }
+            }
+        } finally {
+            readerGroupManager.deleteReaderGroup(readerGroup);
```
I don't think we necessarily want to delete the group because one reader shutdown.

claudiofahey replied:
In this case we do want to delete the reader group. Each instance of EventDebugSink will read the entire stream so it creates a random reader group for itself and then it cleans up when it is done. There's no need for load balancing between EventDebugSink instances (in contrast to AtLeastOnceApp).

claudiofahey replied:
I also added a `createStream` private method to this class.

tkaitchuck commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventDebugSink.java:
```diff
@@ -0,0 +1,98 @@
+        final ReaderGroupConfig readerGroupConfig = ReaderGroupConfig.builder()
+                .stream(Stream.of(getConfig().getScope(), getConfig().getStream2Name()))
+                .build();
+        @Cleanup
```
You can just call close. People shouldn't need to lookup how lombok works to understand our examples.

claudiofahey replied:
I changed this to try-with-resources.

tkaitchuck commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventGenerator.java:
```diff
@@ -0,0 +1,94 @@
+        }
+
+        final Random rand = new Random(42);
+        @Cleanup
```
These two should probably be try-with resources.

claudiofahey replied:
Done.

claudiofahey replied:
I added a `createStreams` private method to this class.

tkaitchuck commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventGenerator.java:
```diff
@@ -0,0 +1,94 @@
+        final SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX");
+        for (;;) {
+            sequenceNumber++;
+            final SampleEvent event = new SampleEvent();
```
Let's make sample event immutable. It can do all this logic in its constructor to make this code neater.

claudiofahey replied:
Done. It is a little awkward because it is using the SampleEvent class as both the input and the output of the transformation that SampleEventProcessor performs.

tkaitchuck commented on pravega-client-examples/src/test/java/io/pravega/example/streamprocessing/WorkerProcessGroup.java:
```diff
@@ -0,0 +1,81 @@
+ * This manages a group of WorkerProcess instances.
+ */
+@Builder
+public class WorkerProcessGroup implements AutoCloseable {
```
The value of this class is not clear to a casual reader.

claudiofahey replied:
I added a comment.
```
 * This manages a group of WorkerProcess instances.
 * The testing framework in {@link StreamProcessingTest} uses this class to
 * start and stop multiple instances of {@link AtLeastOnceProcessorInstrumented}
 * in a single process.
```

tkaitchuck commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventGenerator.java:
```diff
@@ -0,0 +1,95 @@
+        createStreams();
+        final Random rand = new Random(42);
+        try (final EventStreamClientFactory clientFactory = EventStreamClientFactory.withScope(getConfig().getScope(), clientConfig)) {
+            try (final EventStreamWriter<SampleEvent> writer = clientFactory.createEventWriter(
```
These don't need to nest. These two items can be in the same try block.

tkaitchuck commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/EventGenerator.java:
```diff
@@ -0,0 +1,95 @@
+                long sum = 0;
+                for (; ; ) {
+                    sequenceNumber++;
+                    final String routingKey = String.format("%3d", rand.nextInt(1000));
```
Random locals like this don't need to be declared final.

tkaitchuck commented on pravega-client-examples/src/test/java/io/pravega/example/streamprocessing/StreamProcessingTest.java:
```diff
@@ -0,0 +1,544 @@
+public class StreamProcessingTest {
+    static final Logger log = LoggerFactory.getLogger(StreamProcessingTest.class);
+
+    protected static final AtomicReference<SetupUtils> SETUP_UTILS = new AtomicReference<>();
```
This is bad from. Please use a resource to avoid this pattern.

RaulGracia commented:
@claudiofahey can you address @tkaitchuck comments so we can close this one? I think we are close.

shrids commented on pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/AtLeastOnceProcessor.java:
```diff
@@ -0,0 +1,113 @@
+            Position lastFlushedPosition = null;
+            try {
+                while (isRunning()) {
+                    final EventRead<T> eventRead = reader.readNextEvent(readTimeoutMillis);
```
`readNextEvent` can throw a `TruncatedDataException` when StreamRetention kicks in. We would need to retry reading the event in such a scenario, right?

RaulGracia commented:
@claudiofahey any updates addressing @tkaitchuck feedback? This one is actually looks close to get merged, wondering if it is worth it to do a last push for it.
